tsdb/engine/tsm1/encoding.go

```diff
@@ -0,0 +1,571 @@
+	Value() interface{}
+	Size() int
+}
+
+func NewValue(t time.Time, value interface{}) Value {
+	switch v := value.(type) {
+	case int64:
+		return &Int64Value{time: t, value: v}
+	case float64:
+		return &FloatValue{time: t, value: v}
+	case bool:
+		return &BoolValue{time: t, value: v}
+	case string:
+		return &StringValue{time: t, value: v}
+	}
+	return &EmptyValue{}
+}
+
+type EmptyValue struct {
+}
+
+func (e *EmptyValue) UnixNano() int64    { return tsdb.EOF }
+func (e *EmptyValue) Time() time.Time    { return time.Unix(0, tsdb.EOF) }
+func (e *EmptyValue) Value() interface{} { return nil }
+func (e *EmptyValue) Size() int          { return 0 }
+
+// Values represented a time ascending sorted collection of Value types.
+// the underlying type should be the same across all values, but the interface
+// makes the code cleaner.
+type Values []Value
+
+func (v Values) MinTime() int64 {
+	return v[0].Time().UnixNano()
+}
+
+func (v Values) MaxTime() int64 {
+	return v[len(v)-1].Time().UnixNano()
+}
+
+func (v Values) Encode(buf []byte) ([]byte, error) {
+	switch v[0].(type) {
+	case *FloatValue:
+		a := make([]*FloatValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*FloatValue)
+		}
+		return encodeFloatBlock(buf, a)
+
+	case *Int64Value:
+		a := make([]*Int64Value, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*Int64Value)
+		}
+		return encodeInt64Block(buf, a)
+
+	case *BoolValue:
+		a := make([]*BoolValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*BoolValue)
+		}
+		return encodeBoolBlock(buf, a)
+
+	case *StringValue:
+		a := make([]*StringValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*StringValue)
+		}
+		return encodeStringBlock(buf, a)
+	}
+
+	return nil, fmt.Errorf("unsupported value type %T", v[0])
+}
+
+func (v Values) DecodeSameTypeBlock(block []byte) Values {
+	switch v[0].(type) {
+	case *FloatValue:
+		a, _ := decodeFloatBlock(block)
+		return a
+	case *Int64Value:
+		a, _ := decodeInt64Block(block)
+		return a
+	case *BoolValue:
+		a, _ := decodeBoolBlock(block)
+		return a
+	case *StringValue:
+		a, _ := decodeStringBlock(block)
+		return a
+	}
+	return nil
+}
+
+// DecodeBlock takes a byte array and will decode into values of the appropriate type
+// based on the block
+func DecodeBlock(block []byte) (Values, error) {
+	if len(block) == 0 {
+		return Values{}, nil
+	}
+
+	blockType := block[8]
+	switch blockType {
+	case BlockFloat64:
+		return decodeFloatBlock(block)
+	case BlockInt64:
+		return decodeInt64Block(block)
+	case BlockBool:
+		return decodeBoolBlock(block)
+	case BlockString:
+		return decodeStringBlock(block)
+	default:
+	}
+
+	// TODO: add support for other block types
+	return nil, fmt.Errorf("unknown block type: %d", blockType)
+}
+
+// Deduplicate returns a new Values slice with any values
+// that have the same  timestamp removed. The Value that appears
+// last in the slice is the one that is kept. The returned slice is in ascending order
+func (v Values) Deduplicate() Values {
```

Comment:
Are values always in time order (I thought they were).  If so, can't we loop through, track the last time, and if we see it, just keep slicing out the value we don't want?  This would be more performant.